### DISCUSSION TITLE
Adonis 5: Best way to implement traits?

### DISCUSSION THREAD

**Discussion by timboozle** [adonisjs/core#1402] [Help] [answered]
What would be the best way to implement traits in v5?

I have a model `Feed` that has some relatively complex operations I used to use traits for in v4. 
For example `Feed.download()` and `Feed.parse()`

**Comment by LeCoupa** [comment, marked as answer]
Create a class that extends `BaseModel` and inherit this new class?

**Reply by hariprasath-yadav** [reply]
Thank You very much it works for me

**Reply by hariprasath-yadav** [reply]
Yaa its worked at first, but later if I try to use the trait on many models it gives me unexpected malfunctions in my project,
I don't get why these problems are happening but values between modules are mixing with each other so I revert the programs to original as previous. so now I don't recommend this solution.

**Reply by LeCoupa** [reply]
Can you show us your code?

**Reply by Banashek** [reply]
I'd also be curious how you'd go about implementing multiple traits as well.
Inheritance will have a limitation of only allowing a single "trait" class to be implemented.

In rails, mixins allow you to add lots of abstracted features to models, like `Archivable`, `Avatarable`, `Joinable`, etc that are almost like implemented interfaces that hook into the lifecycle of the ActiveRecord model operations (and into the migrations as well to add required fields)

**Reply by hariprasath-yadav** [reply]
Surely i will give you all the answers but right now I'm on the deadline to submit my project, so i will replay you as soon as possible
@Banashek for you check **Poelinca Dorin** answer in the below given link
[ES6 Class Multiple inheritance](https://stackoverflow.com/questions/29879267/es6-class-multiple-inheritance)

**Reply by thetutlage** [reply]
It's not only about AdonisJS. Its about any code written in Typescript. When using traits or mixins, you have to make sure that the new properties are also picked up by the type system.

**Reply by hsharma1996** [reply]
@thetutlage any solution form your side, any suggestion on sharing common function between models ? The extend class method does not work. Is there any other way we can achieve it ?

**Reply by thetutlage** [reply]
Shared https://github.com/adonisjs/core/discussions/3188

**Reply by RosarioDeveloper** [reply]
Does someone already solve this problem?

I have got the same problem and I would like it if someone has already solved this problema.

**Reply by RomainLanz** [reply]
You can have a look at @thetutlage answer above yours.

**Comment by alexandreguertin** [comment]
@timboozle 

I'm not familiar with Typescript. All feedback will be taken into account and greatly valued.

Thanks to [Adonis Notifications](https://github.com/verful/notifications) mixins for the inspiration, here's a working example of a "Trait" that adds translation to a model.

```typescript
// App/Models/Mixins/TranslationMixin.ts

import { NormalizeConstructor } from '@ioc:Adonis/Core/Helpers'
import { BaseModel, column, HasMany, hasMany } from '@ioc:Adonis/Lucid/Orm'

/**
 * Generates the translation model used in the relation. (e.g. ConditionTranslation)
 */
function TranslationModelFactory (tableName: string, columnNames: string[]) {
  class TranslationModel extends BaseModel {
    public static table = tableName

    @column({ isPrimary: true })
    public id: number

    @column()
    public language: string
  }

  /**
   * Adds decorator "column" to the defined columns and foreign key.
   */
  for (const columnName of columnNames) {
    TranslationModel.prototype[columnName] = null
    column()(TranslationModel.prototype, columnName)
  }

  return TranslationModel
}

export default function (modelName: string, columnNames: string[], options?: {
  tableName?: string,
  localKey?: string,
  foreignKey?: string,
}) {
  return function<T extends NormalizeConstructor<typeof BaseModel>> (Base: T) {
    const opts = {
      tableName: `${modelName.toLowerCase()}_translations`,
      localKey: 'id',
      foreignKey: `${modelName.toLocaleLowerCase()}_id`,
      ...options,
    }

    const TranslationModel = TranslationModelFactory(opts.tableName, [opts.foreignKey, ...columnNames])

    /**
     * The actual "Trait".
     */
    class TranslationMixin extends Base {
      /**
       * Because the relationship requires it, we're duplicating the column declaration here.
       */
      @column({ isPrimary: true })
      public id: any

      @hasMany(() => TranslationModel, {
        localKey: opts.localKey,
        foreignKey: opts.foreignKey,
      })
      public translations: HasMany<typeof TranslationModel>
    }

    return TranslationMixin
  }
}
```

```typescript
// App/Models/Condition.ts

import { DateTime } from 'luxon'
import { compose } from '@ioc:Adonis/Core/Helpers'
import { BaseModel, column } from '@ioc:Adonis/Lucid/Orm'

import TranslationMixin from './Mixins/TranslationMixin'

export default class Condition extends compose(
  BaseModel,
  TranslationMixin('Condition', ['title'])
) {
  @column({ isPrimary: true })
  public id: number

  @column.dateTime({ autoCreate: true })
  public createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  public updatedAt: DateTime
}
```
